### PR TITLE
Enable network BPF modules only when needed.

### LINKED ISSUE #17133: Restricted sessions modules causes problems in 5.14 Kernel
Expected behavior:

Enhanced logging will work even if restricted session modules aren't available. If restricted sessions isn't required it shouldn't attempt to load those modules on that machine.

Current behavior:

A 5.14 kernel SUSE fips will fail to run when Teleport attempts to load modules for restricted sessions.

Bug details:
- Teleport version: 10.2.2
- Recreation steps

  Install a FIPS enabled Suse 15 vm that uses 5.14 as the default Kernel. Attempt to enable enhanced logging.  This will not occur in a non-FIPS env. 

- Debug logs

```
[  192.420549][ T5037] BUG: unable to handle page fault for address: ffffffffc13f7bcc
[  192.423075][ T5037] #PF: supervisor instruction fetch in kernel mode
[  192.424999][ T5037] #PF: error_code(0x0010) - not-present page
[  192.426955][ T5037] PGD 1d815067 P4D 1d815067 PUD 1d817067 PMD 0 
[  192.513528][ T5037] Oops: 0010 [#1] PREEMPT SMP NOPTI
[  192.515296][ T5037] CPU: 0 PID: 5037 Comm: teleport Kdump: loaded Tainted: G           OE  X    5.14.21-150400.24.18-default #1 SLE15-SP4 695ab7a8fc20f5ddb345280570966cd1eb06d469
```

### PR DESCRIPTION
**Pull request by jakule**
Currently, all BPF modules are loaded even when they are not explicitly required. This change splits the modules between enhanced and restricted sessions. Now the restricted session modules won't be loaded unless explicitly requested in the configuration. The restricted session configuration struct has been moved to avoid circular dependency.
I also fixed the restricted session test.

Closes https://github.com/gravitational/teleport/issues/17133

### CODE REVIEW

File: lib/bpf/bpf.go [hunk 1]
```diff
@@ -195,12 +206,16 @@ func (s *Service) Close() error {
...
 	s.open.close()
-	s.conn.close()
+	if s.conn != nil {
+		s.conn.close()
+	}
 
 	// Close cgroup service.
-	s.cgroup.Close()
+	if err := s.cgroup.Close(); err != nil {
+		log.Warnf("Failed to close cgroup: %v", err)
```
Comment (codingllama) [inline lib/bpf/bpf.go:215]: Suggestion:

```suggestion
		log.WithError(err).Warn("Failed to close cgroup")
```

File: lib/bpf/common.go [hunk 2]
```diff
@@ -117,6 +117,26 @@ func (c *Config) CheckAndSetDefaults() error {
 	return nil
 }
 
+// RestrictedSessConfig holds configuration for the RestrictedSession service.
+type RestrictedSessConfig struct {
```
Comment (codingllama) [inline lib/bpf/common.go:121]: Spell it out? It's part of the public API.

```suggestion
// RestrictedSessionsConfig holds configuration for the RestrictedSession
// service.
type RestrictedSessionsConfig struct {
```

File: lib/restrictedsession/restricted_test.go [hunk 3]
```diff
@@ -30,7 +30,7 @@ import (
 	"testing"
 	"time"
 
-	go_cmp "github.com/google/go-cmp/cmp"
+	gocmp "github.com/google/go-cmp/cmp"
```
Comment (codingllama) [inline lib/restrictedsession/restricted_test.go:33]: Just call it `cmp`? I don't think the alias is necessary.

```suggestion
	"github.com/google/go-cmp/cmp"
```

File: lib/restrictedsession/restricted_test.go [hunk 4]
```diff
@@ -152,20 +149,32 @@ func setupBPFContext(t *testing.T) *bpfContext {
 		t.Skip("Tests for package restrictedsession can only be run as root.")
 	}
 
+	// Create temporary directory where cgroup2 hierarchy will be mounted.
+	// DO NOT MOVE THIS LINE.
+	// t.TempDir() creates t.Cleanup() action. If this hook is called before
+	// we mount cgroup in Close() this test will fail.
+	cgroupDir := t.TempDir()
+
+	tt := bpfContext{
```
Comment (codingllama) [inline lib/restrictedsession/restricted_test.go:158]:
```suggestion
	bpfCtx := bpfContext{
```

`tt` is typically a test case, which this doesn't seem to be.

File: lib/bpf/bpf.go [hunk 5]
```diff
@@ -171,19 +171,30 @@ func New(config *Config) (BPF, error) {
...
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		log.Debugf("Started enhanced session recording with buffer sizes (command=%v, "+
+			"disk=%v, network=%v) and cgroup mount path: %v. Took %v.",
+			*s.CommandBufferSize, *s.DiskBufferSize, *s.NetworkBufferSize, s.CgroupPath,
+			time.Since(start))
+
+		go s.processNetworkEvents()
```
Comment (greedy52) [inline lib/bpf/bpf.go:187]: any risk of race conditions in those `s.emitXXX` functions now we have two separate event loops? For example, we do `s.argsCache.Get` then `s.argsCache.Set`, probably ok...

**Review by greedy52 — APPROVED** [review]

File: lib/service/cfg.go [hunk 6]
```diff
@@ -1179,7 +1178,7 @@ type TracingConfig struct {
 	SamplingRate float64
 }
 
-// Config generates a tracing.Config that is populated from the values
+// Config generates a tracing.RestrictedSessConfig that is populated from the values
```
[the remark this answers is not in the document]
Comment (jakule) [reply lib/service/cfg.go:1181]: Yeap, thanks.

File: lib/restrictedsession/restricted_test.go (hunk 3 above)
Comment (jakule) [reply lib/restrictedsession/restricted_test.go:33]: This package already has a method called `cmp` Go doesn't allow you to have a collision. I can rename that function but sounds like the same deal, rename something, so `cmp` can be used in a different place.

Comment (codingllama) [reply lib/restrictedsession/restricted_test.go:33]: It's alright, let's keep the alias.

File: lib/bpf/bpf.go [hunk 7]
```diff
@@ -171,19 +171,30 @@ func New(config *Config) (BPF, error) {
...
-	s.conn, err = startConn(*config.NetworkBufferSize)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
 
-	log.Debugf("Started enhanced session recording with buffer sizes (command=%v, "+
-		"disk=%v, network=%v) and cgroup mount path: %v. Took %v.",
-		*s.CommandBufferSize, *s.DiskBufferSize, *s.NetworkBufferSize, s.CgroupPath,
-		time.Since(start))
+	if restrictedSession.Enabled {
```
[the remark this answers is not in the document]
Comment (jakule) [reply lib/bpf/bpf.go:178]: Added above.

File: lib/bpf/common.go [hunk 8]
```diff
@@ -117,6 +117,26 @@ func (c *Config) CheckAndSetDefaults() error {
...
+	// Enabled if this service will try and install BPF programs on this system.
+	Enabled bool
+
+	// EventsBufferSize is the size (in pages) of the perf buffer for events.
```
[the remark this answers is not in the document]
Comment (codingllama) [reply lib/bpf/common.go:125]: Bump?

**Review by codingllama — APPROVED** [review]

**Pull request merged**